Add status and management UI for the event indexer

poljar opened the pull request:
This PR implements some initial UI for our event indexer.
Things that are configurable/shown:
- [x] Enable/disable the crawler
- [x] Setting controlling how fast the room history is being downloaded.
- [x] See how much disk usage the event index is consuming
- [x] See if the crawler is currently downloading room history.

Some more things  we might want:
- [ ] Configure the event index to crawl non-encrypted rooms as well.
- [ ] Trigger a manual re-crawl of a room
- [ ] A decoration showing the current shown room is currently being crawled.
- [ ] Language settings (some discussion for this can be found already here https://github.com/matrix-org/seshat/issues/7)

As for configuring non-encrypted rooms we should consider if we want to allow users to index any type of room or limit it to non-public rooms or DMs only.

Rendered:

![image](https://user-images.githubusercontent.com/1294269/69711861-b2bf3d00-10f9-11ea-9cf6-54df14f6e6cc.png)

The settings have been crammed into the preferences tab of the user settings. This is most likely the completely wrong place to put them since we already have quite a large number of settings there.

A new tab for the settings was also considered but we may have enough tabs as well.

Please note that this PR depends on changes in Riot-web as well (https://github.com/vector-im/riot-web/pull/11529)

Fixes https://github.com/vector-im/riot-web/issues/11984

ara4n commented:
From a high level functionality perspective this looks great.  I think we can afford to expose a little more information to the user to give them more visibility and a sense of control though (at least until we are confident this works seamlessly).

 * [ ] s/Enable message downloading/Download and index encrypted messages/
 * [ ] The number of indexed messages (570KB in 125,432 messages over 129 rooms; 248KB of index)
 * [ ] Listing the names of the rooms currently being indexed
 * [ ] Delete cache button

In terms of where it should sit in the UI - I'd have expected it to be in Security & Privacy alongside the rest of the E2E settings.

@nadonomy: can you provide a design for this please?

nadonomy commented:
@poljar for when you're back after Christmas/New Years I've added a design proposal for the sestet UI here:

https://www.figma.com/file/wwqBo5oAFfk8XKmtd2YaIs/Cross-signing?node-id=2011%3A497

Let me know when you're back online on Matrix and we can iterate on some of the details. From the requirements I could gather from this issue, there's a few things that would be good to discuss:

- Until we're clear of cross-signing dev, Settings in /develop doesn't represent the designs here, so we should quickly discuss where to place this within the 'Security & Privacy' tab in settings
- I'm not sure that users have a good expectation of what a [Delete cache] button would actually do— e.g. would this automatically start downloading & unencrypting messages again? Or would that be user initiated? So I think it makes more sense to keep the interaction to [Disable] & [Enable] the feature, where doing so would clear the cache.
- I'm not convinced of the utility of listing the names of the rooms messages are being fetched from; as the user can't manage any of it and it's problematic to scale if e.g. the user is in 1000 encrypted rooms.
- Are we implementing this on iOS/Android?

poljar commented:
> Until we're clear of cross-signing dev, Settings in /develop doesn't represent the designs here, so we should quickly discuss where to place this within the 'Security & Privacy' tab in Settings in the meanwhile

I guess bellow the key backup section seems sensible?

> I'm not sure that users have a good expectation of what a [Delete cache] button would actually do— e.g. would this automatically start downloading & unencrypting messages again? Or would that be user initiated? So I think it makes more sense to keep the interaction to [Disable] & [Enable] the feature, where doing so would clear the cache.
> 

As it is, it would restart the download. Agree a Disable/Enable could be enough.

>  I'm not convinced of the utility of listing the names of the rooms messages are being fetched from; as the user can't manage any of it and it's problematic to scale if e.g. the user is in 1000 encrypted rooms.

Agree, a summary should be enough, but some type of marker could be nice for when a user looks at a specific room. Just some visual hint if messages are being downloaded or not.

> Are we implementing this on iOS/Android?

Eventually, yes. No progress as of yet on that side though.

poljar commented:
Some progress here, refactored the main UI, moved it into the security tab of the settings and moved it into a modal.

I force pushed this after i locally rebased onto develop since the original PR was done on the old build system.

The disabling/enabling isn't yet hooked up.

Looks like this now:

![image](https://user-images.githubusercontent.com/552026/72805582-8c2e0600-3c53-11ea-9918-8e2d09f6825e.png)

poljar commented:
Updated to reflect the suggested design in figma. The buttons may need some more love, or rather less since they are too close together.

Looks like this
![image](https://user-images.githubusercontent.com/552026/72995992-26c84980-3dfa-11ea-9f6d-217f8af4c775.png)

Video of the whole interaction can be found [here](https://streamable.com/1evwx#).

jryans commented on src/async-components/views/dialogs/eventindex/ManageEventIndex.js:
```diff
@@ -0,0 +1,160 @@
+                }
+                <div className='mx_SettingsTab_subsectionText'>
+                    {_t("Space used:")} {formatBytes(this.state.eventIndexSize, 0)}<br />
+                    {_t("Indexed messages:")} {this.state.eventCount}<br />
```
For localised numbers, first stash a number formatter somewhere (this component or `FormattingUtils` perhaps):

```js
const formatNumber = new Intl.NumberFormat().format;
```

and then it should work for rooms and events:

```js
                    {_t("Indexed messages:")} {formatNumber(this.state.eventCount)}<br />
                    {_t("Number of rooms:")} {formatNumber(this.state.roomCount)}<br />
```

poljar replied:
I used `formatCount` in `FormattingUtils` since it seems to be used through the code-base.

jryans commented on src/components/views/settings/EventIndexPanel.js:
```diff
@@ -0,0 +1,158 @@
+                        {_t( "Securely cache encrypted messages locally for them " +
+                             "to appear in search results, using ")
+                        } {formatBytes(this.state.eventIndexSize, 0)}
+                        {_t( " to store messages from ")} {this.state.roomCount} {_t("rooms.")}
```
Use number format again here.

jryans commented on src/components/views/settings/EventIndexPanel.js:
```diff
@@ -0,0 +1,158 @@
+                <div>
+                    {
+                        _t( "Riot can't securely cache encrypted messages locally " +
+                            "while running in a web browser. Use Riot Desktop for " +
```
I think the design intends for this to be a link to the Riot Desktop download page. Searching the code for `=> <a` should reveal many examples of how to add a link in a translated string. I assume it should link to https://riot.im/download/desktop.

poljar replied:
Added, but on that note, don't we want to show something different in the case that the platform supports event indexing but support isn't installed.

The case when we're under Riot Desktop but Seshat is missing.

jryans replied:
Right, good point! Let's add a separate variant for the case where we are on Electron but Seshat is missing then, with some text and link to docs like:

> Riot is missing some components required for securely caching encrypted messages locally. If you'd like to experiment with this feature, build a custom Riot Desktop with [search components added](https://github.com/vector-im/riot-web/blob/develop/docs/native-node-modules.md#adding-seshat-for-search-in-e2e-encrypted-rooms).

poljar replied:
Thanks, added.

jryans commented on src/indexing/EventIndex.js:
```diff
@@ -64,59 +69,61 @@ export default class EventIndex {
     }
 
-    onSync = async (state, prevState, data) => {
+    /** Get crawler checkpoints for the encrypted rooms and store them in the index.
```
We typically nudge the text to the next line like

```
/**
 * text
 */
```

jryans commented on src/indexing/EventIndex.js:
```diff
@@ -64,59 +69,61 @@ export default class EventIndex {
-                const isRoomEncrypted = (room) => {
-                    return client.isRoomEncrypted(room.roomId);
-                };
+        // We only care to crawl the encrypted rooms, non-encrypted.
```
```suggestion
        // We only care to crawl the encrypted rooms, non-encrypted
```

jryans commented on src/indexing/EventIndex.js:
```diff
@@ -406,4 +433,37 @@ export default class EventIndex {
+        }
+    }
+
+    on(eventName, callback) {
```
You can also have your class extend `EventEmitter` if you like to avoid this, but it's up to you.

jryans commented on src/indexing/EventIndexPeg.js:
```diff
@@ -40,12 +42,32 @@ class EventIndexPeg {
+        return this.initEventIndex();
+    }
+
+    /* Initialize the event index.
```
Two stars, move to next line.

poljar commented on src/async-components/views/dialogs/eventindex/DisableEventIndex.js:
```diff
@@ -0,0 +1,78 @@
+ * Walks the user through the process of creating an e2e key backup
+ * on the server.
+ */
+export default class ManageEventIndex extends React.Component {
```
I feel like we already know that it's a dialog since it's in the dialogs folder, but added.

jryans replied:
Well, maybe, but better to be consistent. Happy to discuss a separate PR that removes the suffix from all dialogs if you feel strongly about it. 😄

poljar replied:
I think it's fine. 😅

Maybe some other time.

poljar commented on src/async-components/views/dialogs/eventindex/ManageEventIndex.js:
```diff
@@ -0,0 +1,160 @@
+            </div>
+        );
+
+        const buttons = (
```
I believe I can't make the left button (the cancel one) marked as `danger` using `DialogButtons`, should I make that possible or did you have something else in mind.

jryans replied:
Yes, looks like you're right. We could expose a new prop through `DialogButtons` to support a class name for the cancel button, or else stick with this. If you keep this approach, we'll need a bit of CSS to add the spacing.

jryans commented on src/async-components/views/dialogs/eventindex/DisableEventIndexDialog.js:
```diff
@@ -55,23 +52,24 @@ export default class ManageEventIndex extends React.Component {
+                    primaryButtonClass="danger"
+                    onCancel={this.props.onFinished}
+                    disabled={this.state.disabling}
+                >
```
If you're keeping the spinner outside the button row like this, and thus there's no children for `DialogButtons`, you can use a self-closing tag instead here.

poljar replied:
I tried to keep the spinner inline, but the buttons would then be one above the other. It seems that `DialogButtons` doesn't expect any content. Will remove the separate closing tag.

jryans commented on src/async-components/views/dialogs/eventindex/ManageEventIndexDialog.js:
```diff
@@ -20,16 +20,14 @@ import PropTypes from 'prop-types';
 
 import Modal from '../../../../Modal';
-import {formatBytes} from "../../../../utils/FormattingUtils";
+import {formatBytes, formatCount} from "../../../../utils/FormattingUtils";
```
Hmm, so `formatCount` is designed for cramming into tiny spaces like a badge icon. For this PR, we have more space to show these numbers, so let's add a new formatting util to show the full number. I think it will give a better sense of progress that way, as you can still see numbers change regularly once you pass 1k, etc.

jryans commented on src/async-components/views/dialogs/eventindex/DisableEventIndexDialog.js:
```diff
@@ -0,0 +1,76 @@
+        const DialogButtons = sdk.getComponent('views.elements.DialogButtons');
+
+        return (
+            <BaseDialog className='mx_ManageEventIndexDialog'
```
This is the class name for the other dialog... but probably you can just remove it entirely since it's not currently used to style anything at the moment.

jryans commented on src/async-components/views/dialogs/eventindex/DisableEventIndexDialog.js:
```diff
@@ -0,0 +1,76 @@
+            >
+                {_t("If disabled, messages from encrypted rooms won't appear in search results.")}
+                {this.state.disabling ? <Spinner /> : <div />}
+                <DialogButtons
```
For this one, where we want something like:

<img width="200" alt="2020-01-24 at 13 34" src="https://user-images.githubusercontent.com/279572/73073011-4de05300-3eae-11ea-881e-b540a557f0a1.png">

I don't think we have a standard easy way to style the "cancel" button with red text and no background. Here's [an approach](https://github.com/matrix-org/matrix-react-sdk/blob/7de5f09f9f1844fa9fc0d1a84dcf7671aac3cfba/src/components/structures/auth/CompleteSecurity.js#L143-L157) I used recently to achieve it. Since we don't have `DialogButtons` with that approach, you'll a little CSS to space out the buttons, such as [this block](https://github.com/matrix-org/matrix-react-sdk/blob/7de5f09f9f1844fa9fc0d1a84dcf7671aac3cfba/res/css/structures/auth/_CompleteSecurity.scss#L40-L51).

(We should generalised this stuff it seems, but maybe we'll wait for it to appear a few more times first.)

poljar replied:
Can't we just add a `warning` class and appropriate style for that class? This way I can use the prop that I added to the `DialogButtons`

jryans replied:
Yes, feel free to use a `warning` class for this style. Adding some more CSS [near the dialog button styles](https://github.com/matrix-org/matrix-react-sdk/blob/develop/res/css/_common.scss#L425-L429) is probably what would be needed.

poljar commented:
Updated now to have a red transparent cancel button on the disable dialog.

I left the border around the button since it seemed a bit weird not having it there. The colors aren't the transparent ones like in the figma but I assume we want to switch those around all at once and be consistent for now.

Updated screenshots follow:
![image](https://user-images.githubusercontent.com/552026/73183277-d5ca9500-411a-11ea-9ed4-8a428c0738fb.png)

![image](https://user-images.githubusercontent.com/552026/73183174-b469a900-411a-11ea-9c74-3f34bc634ce7.png)

![image](https://user-images.githubusercontent.com/552026/73183314-e418b100-411a-11ea-897c-21880b39df51.png)

jryans commented on src/components/views/settings/EventIndexPanel.js:
```diff
@@ -140,6 +141,28 @@ export default class EventIndexPanel extends React.Component {
+                            "with <nativeLink>search components added</nativeLink>.",
+                            {},
+                            {
+                                'nativeLink': (sub) => <a href={nativeLink}>{sub}</a>,
```
Please add `target="_blank" rel="noopener"` to this link and the one below so they open in new tabs.
